File: modules/fbs-core/api/src/main/kotlin/de/thm/ii/fbs/services/v2/persistence/storage/FsStorageService.kt
package de.thm.ii.fbs.services.v2.persistence.storage

import de.thm.ii.fbs.model.v2.storageBucketName
import de.thm.ii.fbs.model.v2.storageFileName
import de.thm.ii.fbs.utils.v2.exceptions.NotFoundException
import org.apache.commons.io.FileUtils
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import java.util.Optional
import kotlin.io.path.deleteIfExists

@Component
class FsStorageService {
    @Value("\$storage.uploadDir")
    lateinit var uploadDir: String

    private fun uploadDirPath(): Path = Path.of(uploadDir)

    private fun tasksDir(tid: Int) = uploadDirPath().resolve(storageBucketName.CHECKER_CONFIGURATION_FOLDER).resolve(tid.toString())

    private fun submissionDir(sid: Int) = uploadDirPath().resolve(storageBucketName.SUBMISSIONS_BUCKET).resolve(sid.toString())

    fun getFileContent(path: Optional<Path>): String {
        return if (path.isPresent) {
            path.get().toFile().bufferedReader().use { it.readText() }
        } else {
            ""
        }
    }

    @Throws(IOException::class)
    fun storeConfigurationFile(tid: Int, src: Path, fileName: String) =
            Files.move(src, Files.createDirectories(tasksDir(tid)).resolve(fileName), StandardCopyOption.REPLACE_EXISTING)

    /**
     * Store (replace if exists) the solution file a submission
     *
     * @param sid Submission id
     * @param src Current path to the file
     * @throws IOException If the i/o operation fails
     */
    @Throws(IOException::class)
    fun storeSolutionFile(sid: Int, src: Path) =
            Files.move(src, Files.createDirectories(submissionDir(sid)).resolve(storageFileName.SOLUTION_FILE), StandardCopyOption.REPLACE_EXISTING)

    /**
     * Get the path to the main file of a task
     *
     * @param tid Task id
     * @return The path to the file
     */
    fun pathToMainFile(tid: Int): Optional<Path> = Optional.of(tasksDir(tid).resolve(storageFileName.MAIN_FILE))

    /**
     * Get the path to the secondary file of a task
     *
     * @param tid Task id
     * @return The path to the file
     */
    fun pathToSecondaryFile(tid: Int): Optional<Path> = Optional.of(tasksDir(tid).resolve(storageFileName.SECONDARY_FILE))

    /**
     * Get the path to the solution file of a submission
     *
     * @param sid Submission id
     * @return The path to the file
     */
    fun pathToSolutionFile(sid: Int): Optional<Path> = Optional.of(submissionDir(sid).resolve(storageFileName.SOLUTION_FILE))

    /**
     * Gets the Content of the solution file
     *
     * @param sid Submission id
     * @return The Solution file content
     */
    fun getSolutionFile(sid: Int): String = getFileContent(pathToSolutionFile(sid))

    /**
     * Gets the Content of the main file
     *
     * @param ccid Checkrunner id
     * @return The Solution file content
     */
    fun getMainFile(ccid: Int): String = getFileContent(pathToMainFile(ccid))

    /**
     * Gets the Content of the secondary file
     *
     * @param ccid Checkrunner id
     * @return The Solution file content
     */
    fun getSecondaryFile(ccid: Int): String = getFileContent(pathToSecondaryFile(ccid))

    /**
     * Delete a main file
     *
     * @param tid Task id
     * @return True if deteled, false if not file exists
     * @throws IOException If the i/o operation fails
     */
    @Throws(IOException::class)
    fun deleteMainFile(tid: Int): Boolean {
        val tmp = pathToMainFile(tid)
        if (tmp.isPresent) {
            return tmp.get().deleteIfExists()
        }
        throw NotFoundException()
    }

    /**
     * Delete a secondary file
     *
     * @param tid Task id
     * @return True if deteled, false if not file exists
     * @throws IOException If the i/o operation fails
     */
    @Throws(IOException::class)
    fun deleteSecondaryFile(tid: Int): Boolean {
        val tmp = pathToSecondaryFile(tid)
        if (tmp.isPresent) {
            return tmp.get().deleteIfExists()
        }
        throw NotFoundException()
    }

    /**
     * Delete the Configuration Folder with all Files inside
     *
     * @param tid Task id
     * @return True if deteled, false if not Directory exists
     * @throws IOException If the i/o operation fails
     */
    fun deleteConfiguration(tid: Int): Boolean
    {
        return deleteFolder(tasksDir(tid))
    }

    /**
     * Delete a solution file
     *
     * @param sid Submission id
     * @return True if deteled, false if not file exists
     * @throws IOException If the i/o operation fails
     */
    @Throws(IOException::class)
    fun deleteSolutionFile(sid: Int): Boolean = deleteFolder(submissionDir(sid))

    private fun deleteFolder(path: Path): Boolean {
        val confDir = path.toFile()

        return if (confDir.exists() && confDir.isDirectory) {
            FileUtils.deleteDirectory(confDir)
            true
        } else {
            false
        }
    }
}

File: modules/fbs-core/api/src/main/scala/de/thm/ii/fbs/services/persistence/storage/FsStorageService.scala
package de.thm.ii.fbs.services.persistence.storage

import de.thm.ii.fbs.model.{storageBucketName, storageFileName}
import org.apache.commons.io.FileUtils
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component

import java.io.{File, FileOutputStream, IOException}
import java.nio.file.{Files, Path, StandardCopyOption}
import scala.io.Source

@Component
class FsStorageService {
  def openClone(path: Path): File = {
    val tmpFile = File.createTempFile("fs-tmp-", null)
    Files.copy(path, new FileOutputStream(tmpFile))
    tmpFile
  }

  @Value("${storage.uploadDir}")
  private val uploadDir: String = null

  private def uploadDirPath: Path = Path.of(uploadDir)

  private def tasksDir(tid: Int) = uploadDirPath.resolve(storageBucketName.CHECKER_CONFIGURATION_FOLDER).resolve(String.valueOf(tid))

  private def submissionDir(sid: Int) = uploadDirPath.resolve(storageBucketName.SUBMISSIONS_BUCKET).resolve(String.valueOf(sid))

  def getFileContent(path: Option[Path]): String = {
    if (path.isDefined) {
      val source = Source.fromFile(path.get.toFile)

      try {
        source.mkString
      } finally {
        source.close()
      }
    } else {
      ""
    }
  }

  @throws[IOException]
  def storeConfigurationFile(tid: Int, src: Path, fileName: String): Unit =
    Files.move(src, Files.createDirectories(tasksDir(tid)).resolve(fileName), StandardCopyOption.REPLACE_EXISTING)

  /**
    * Store (replace if exists) the solution file a submission
    *
    * @param sid Submission id
    * @param src Current path to the file
    * @throws IOException If the i/o operation fails
    */
  @throws[IOException]
  def storeSolutionFile(sid: Int, src: Path): Unit =
    Files.move(src, Files.createDirectories(submissionDir(sid)).resolve(storageFileName.SOLUTION_FILE), StandardCopyOption.REPLACE_EXISTING)

  /**
    * Get the path to the main file of a task
    *
    * @param ccid Checker Configuration id
    * @return The path to the file
    */
  def pathToMainFile(ccid: Int): Option[Path] = Option(tasksDir(ccid).resolve(storageFileName.MAIN_FILE)).filter(Files.exists(_))

  /**
    * Get the path to the secondary file of a task
    *
    * @param ccid Checker Configuration id
    * @return The path to the file
    */
  def pathToSecondaryFile(ccid: Int): Option[Path] = Option(tasksDir(ccid).resolve(storageFileName.SECONDARY_FILE)).filter(Files.exists(_))

  /**
    * Get the path to the solution file of a submission
    *
    * @param sid Submission id
    * @return The path to the file
    */
  def pathToSolutionFile(sid: Int): Option[Path] = Option(submissionDir(sid).resolve(storageFileName.SOLUTION_FILE)).filter(Files.exists(_))

  /**
    * Gets the Content of the solution file
    *
    * @param sid Submission id
    * @return The Solution file content
    */
  def getSolutionFile(sid: Int): String = getFileContent(pathToSolutionFile(sid))

  /**
    * Gets the Content of the main file
    *
    * @param ccid Checkrunner id
    * @return The Solution file content
    */
  def getMainFile(ccid: Int): String = getFileContent(pathToMainFile(ccid))

  /**
    * Gets the Content of the secondary file
    *
    * @param ccid Checkrunner id
    * @return The Solution file content
    */
  def getSecondaryFile(ccid: Int): String = getFileContent(pathToSecondaryFile(ccid))

  /**
    * Delete a main file
    *
    * @param tid Task id
    * @return True if deteled, false if not file exists
    * @throws IOException If the i/o operation fails
    */
  @throws[IOException]
  def deleteMainFile(tid: Int): Boolean = pathToMainFile(tid).exists(Files.deleteIfExists)

  /**
    * Delete a secondary file
    *
    * @param tid Task id
    * @return True if deteled, false if not file exists
    * @throws IOException If the i/o operation fails
    */
  @throws[IOException]
  def deleteSecondaryFile(tid: Int): Boolean = pathToSecondaryFile(tid).exists(Files.deleteIfExists)

  /**
    * Delete the Configuration Folder with all Files inside
    *
    * @param tid Task id
    * @return True if deteled, false if not Directory exists
    * @throws IOException If the i/o operation fails
    */
  def deleteConfiguration(tid: Int): Boolean = {
    deleteFolder(tasksDir(tid))
  }

  /**
    * Delete a solution file
    *
    * @param sid Submission id
    * @return True if deteled, false if not file exists
    * @throws IOException If the i/o operation fails
    */
  @throws[IOException]
  def deleteSolutionFile(sid: Int): Boolean = deleteFolder(submissionDir(sid))

  private def deleteFolder(path: Path) = {
    val confDir = path.toFile

    if (confDir.exists() && confDir.isDirectory) {
      FileUtils.deleteDirectory(confDir)
      true
    } else {
      false
    }
  }
}
